xformers/benchmarks/benchmark_blocksparse_transformers.py

```diff
@@ -0,0 +1,1042 @@
+    block_size: int = 64
+
+    @property
+    def blocked_seq_length(self):
+        return int(self.seq_length / self.block_size)
+
+    def init(self, kwargs):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def __str__(self):
+        return f"bs={self.batch_size},h={self.num_heads},k={self.hidden_size},seq={self.seq_length},bl={self.block_size}"
+        # return "-".join([str(getattr(self, x.name)) for x in fields(self)])
+
+
+class AttentionMask(object):
+    def __init__(self, config=None):
+        super().__init__()
+        if config is None:
+            config = Configuration()
+        self.config = config
+
+    def is_blocked(self):
+        return self.config.block_size != 1
+
+    def is_valid_config(self, keep_blocked=True):
+        return True
+
+    def expand(self, mask):
+        if mask.ndim == 2:
+            return mask.unsqueeze(0).expand(self.config.num_heads, -1, -1)
+
+    def gen_mask(self, keep_blocked=True):
+        raise NotImplementedError("Abstract data class")
+
+    def set_config_attr(self, key, value):
+        setattr(self.config, key, value)
+
+    def __str__(self):
+        raise NotImplementedError("Abstract data type")
+
+    def __call__(self):
+        mask = self.gen_mask()
+        return mask, self.config, str(self)
+
+
+class RandomAttentionMask(AttentionMask):
+    """
+    This is a Random mask. Useful for performance and memory analysis.
+    """
+
+    def __init__(self, config=None):
+        super(RandomAttentionMask, self).__init__(config)
+        self.set_config_attr("mask_prob", 0.5)
+
+    def gen_mask(self, keep_blocked=True):
+        seq_length = self.config.seq_length
+        if keep_blocked:
+            seq_length = self.config.blocked_seq_length
+
+        mask = torch.rand(seq_length, seq_length) > self.config.mask_prob
+        return self.expand(mask)
+
+    def __str__(self):
+        return "random"
+
+
+class LowerTriangularAttentionMask(AttentionMask):
+    """
+    This is a lower triangular mask. This is common in decoder only models.
+    This should reduce the computation and memory to roughly half as close to
+    half of the mask is zero.
+
+    The mask stays same for each head and each input.
+
+    Nit pick (TODO) - While blocking, we need to ensure that the blocks along
+    the diagonals are themselves lower triangular blocks. But, for performance
+    measurement, this is ok to ignore as we treat the whole block as useful
+    values.
+    """
+
+    def __init__(self, config=None):
+        super(LowerTriangularAttentionMask, self).__init__(config)
+
+    def gen_mask(self, keep_blocked=True):
+        seq_length = self.config.seq_length
+        if keep_blocked:
+            seq_length = self.config.blocked_seq_length
+        return self.expand(causal_1d_pattern(seq_length))
+
+    def __str__(self):
+        return "lower_triangular"
+
+
+class BigBirdAttentionMask(AttentionMask):
```

Comment:
@ptillet an example of blocksparse + mask, it's being used by real people I promise :)

Reply:
:upside_down_face: is there any way to define this mask statically?

Reply:
yes actually, I think that it's a very acceptable limitation. In vision there are some approaches to make the attention very dynamic which could benefit from floating masks (handwavy explanation, like RPN for attention), but I think that not supporting that is decent enough.. [Butterfly matrices ](https://dawn.cs.stanford.edu/2019/06/13/butterfly/)is another application but it can be static as far as I remember

Reply:
Yep. Just to be clear, supporting completely dynamic masks is possible, but the overhead is high so I want to incentivize using static attention masks as much as possible. If there is an analytical formula to determine which indices should be omitted, this can be done pretty easily as shown there https://github.com/openai/triton/blob/master/python/triton/ops/blocksparse/softmax.py#L67